### PR TITLE
Change language debug mode to allow it to show the language constants

### PR DESCRIPTION
**Pull request by brianteeman**
PR for #14304

By default there is no change when debug language is enabled - but there now is an option to display either the language constant or the language value

Arguably a new feature - I will let the maintainer decide on that

### CODE REVIEW

**Comment by elkuku** [comment]
I have tested this item :white_check_mark: successfully on 12771998b7e0720c78333f8d9367a21137dcd74f<br /><br />There is an issue that debug settings are not immediately being applied so I guess testers might get confused?

Anyway - works :wink: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21691">issues.joomla.org/tracker/joomla-cms/21691</a>.</sub>

File: libraries/src/Language/Language.php [hunk 1]
```diff
@@ -332,7 +332,9 @@ public function _($string, $jsSafe = false, $interpretBackSlashes = true)
 
 		if (isset($this->strings[$key]))
 		{
-			$string = $this->debug ? '**' . $this->strings[$key] . '**' : $this->strings[$key];
+			$value = $this->debug && \JFactory::getApplication()->get('debug_lang_const') == 0 ? $key : $this->strings[$key];
+
+			$string = $this->debug ? '**' . $value . '**' : $this->strings[$key];
```
Comment (elkuku) [inline libraries/src/Language/Language.php]: I'd propose to refactor this a bit, so the code is only executed when debugging - this should have a measurable impact!

```php
if (isset($this->strings[$key]))
{
	$string = $this->strings[$key];				

	// Store debug information
	if ($this->debug)
	{
		$value = \JFactory::getApplication()->get('debug_lang_const') == 0 ? $key : $string;
		$string = '**' . $value . '**';
				
		$caller = $this->getCallerInfo();
		if (!array_key_exists($key, $this->used))
		{
			$this->used[$key] = array();
		}
		$this->used[$key][] = $caller;
	}
}
```

**Comment by elkuku** [comment]
May I politely ask a question here.. in case this get merged... will the code eventually also end up in the 4.0 branch or does this require a separate PR?

**Comment by brianteeman** [comment]
basically yes it should do as george is pushing them all up (eventually) 

however as this one uses external libraries iirc it might need to be changed so he might ask you to do a specific pr as we handle those differently in j4

**Comment by elkuku** [comment]
> he might ask you to do a specific pr

I'd love to do that.

Thanks!

**Comment by brianteeman** [comment]
@elkuku I have no idea what I was talking about above. You dont have to do anything. If accepted then @wilsonge will eventually merge it into j4 as well. I "think" I replied on my phone and thought you were commenting on your own pr for a new debugger

File: libraries/src/Language/Language.php (hunk 1 above)
Comment (brianteeman) [reply libraries/src/Language/Language.php]: applied your changes

**Comment by elkuku** [comment]
> You dont have to do anything.

I also love to do that :wink: 

Thanks for the update.

**Comment by infograf768** [comment]
Works fine, but the Save toolbar button may be an issue when switching from constant to value.

screen shot 2018-08-25 at 07 45 39

screen shot 2018-08-25 at 07 46 15

and even worse when there is a specific custom admin menu added.

**Comment by brianteeman** [comment]
That is unavoidable if someone wants this feature.

**Comment by elkuku** [comment]
I have tested this item :white_check_mark: successfully on 82fd8edc911d62a281649a2d9759de54a0c93701<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21691">issues.joomla.org/tracker/joomla-cms/21691</a>.</sub>

**Comment by infograf768** [comment]
I suggest to add an onChange event in the form field to save and reload the page.

**Comment by brianteeman** [comment]
That would then be different to every other selector in the ui - so that would not be a good idea

**Comment by tecpromotion** [comment]
I have tested this item :white_check_mark: successfully on 82fd8edc911d62a281649a2d9759de54a0c93701<br /><br />Tested successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21691">issues.joomla.org/tracker/joomla-cms/21691</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by mbabker** [comment]
I'm not going to block the merge on this but we need to update the Language class to not rely on pulling this param from the global config at all times (basically, the same way the class has a parameter to enable/disable debug mode we should have the same for this new capability).

**Comment by brianteeman** [comment]
Thanks for merging. Not sure I can do that change but I will try

**Comment by infograf768** [comment]
This created an issue when debug is on when installing Joomla
See https://github.com/joomla/joomla-cms/issues/28913

**Pull request merged**